processor/resourcedetectionprocessor/internal/aws/eks/detector.go

```diff
@@ -74,9 +82,55 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 	d.rb.SetCloudProvider(conventions.AttributeCloudProviderAWS)
 	d.rb.SetCloudPlatform(conventions.AttributeCloudPlatformAWSEKS)
 
+	clusterName, err := d.getClusterName()
+	if err != nil {
+		return pcommon.NewResource(), "", err
+	}
+	d.rb.SetK8sClusterName(clusterName)
+
 	return d.rb.Emit(), conventions.SchemaURL, nil
 }
 
+func (d *detector) getClusterName() (string, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO())
+	if err != nil {
+		d.logger.Debug("Failed to load default config", zap.Error(err))
+		return "", err
+	}
+
+	ec2IMS := imds.NewFromConfig(cfg)
+	regionOutput, err := ec2IMS.GetRegion(context.TODO(), &imds.GetRegionInput{})
+	if err != nil {
+		d.logger.Debug("Failed to get EC2 instance region", zap.Error(err))
+		return "", err
+	}
+	cfg.Region = regionOutput.Region
+
+	client := ec2.NewFromConfig(cfg)
+	describeInstances, err := client.DescribeInstances(context.TODO(), &ec2.DescribeInstancesInput{})
+	if err != nil {
+		d.logger.Debug("Failed to describe EC2 instances", zap.Error(err))
+		return "", err
+	}
+
+	return d.getClusterNameTagFromReservations(describeInstances.Reservations), nil
+}
+
+func (d *detector) getClusterNameTagFromReservations(reservations []types.Reservation) string {
+	for _, reservation := range reservations {
+		for _, instance := range reservation.Instances {
+			for _, tag := range instance.Tags {
+				if *tag.Key == clusterNameAwsEksTag || *tag.Key == clusterNameEksTag {
+					return *tag.Value
+				}
+			}
+		}
+	}
+
+	d.logger.Debug("Could not find cluster name within EC2 instance tags")
```

Comment:
should this return an error, rather than setting the cluster name to empty?

Reply:
I've decided to just return an empty string, and then create an error within `getClusterName` if it's empty.